inline/src/parser/mod.rs

```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
+            let not_enclosed_and_interrupted = !cxt.enclosed && next_token.kind != kind;
+
+            if enclosed_and_closes {
+                span.end = next_token.span.end;
+                break;
+            } else if not_enclosed_and_interrupted {
+                if next_token.is_plain_whitespace() {
+                    // consume whitespace
+                    let (next_content, next_span) = next_token.parts();
+                    content.push_str(next_content);
+                    span.end = next_span.end;
+
+                    if cxt.merge_whitespace {
+                        // skip other whitespace
+                        while let Some(tkn) = self.next_token() {
+                            if !tkn.is_plain_whitespace() {
+                                self.token_cache = Some(tkn);
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    // cache popped token and break
+                    self.token_cache = Some(next_token);
+                    break;
+                }
+            } else {
+                let (next_content, next_span) = next_token.parts();
+                content.push_str(next_content);
+                span.end = next_span.end;
+            }
         }
 
-        while let Some(next_token) = self.next_token() {
-            end = next_token.span.end;
+        Inline::plain_or_eol(content, span, kind)
+    }
 
-            if next_token.closes() {
+    fn parse_nested(&mut self, start_token: Token) -> Inline {
+        let start = start_token.span.start;
+        let mut end = start_token.span.end;
+        let mut content = VecDeque::new();
+
+        while let Some(next_token) = self.next_token() {
+            if next_token.closes(Some(&start_token)) {
+                end = next_token.span.end;
                 break;
             } else if next_token.opens() {
                 // lexer resolved tokens, if token opens, it is guaranteed that closing exists too.
                 // If not, it's bug in implementation
                 let nested = self.parse_inline(next_token);
-                content.append_inline(nested);
+                content.push_back(nested);
             } else {
-                if kind == TokenKind::Plain && next_token.kind != TokenKind::Plain {
-                    self.token_cache = Some(next_token);
-                    break;
-                }
+                // opening token not yet closed, next token does not start a new inline -> parse as
+                // plain text
 
                 end = next_token.span.end;
-                content.append(InlineContent::from(next_token));
+                let (inner, span) = next_token.parts();
+                let inline = Inline::Plain(Plain {
+                    content: String::from(inner),
+                    span,
+                });
```

Comment:
Oh that's a nice catch 🦅 👀. Generalisation at this point would probably be the correct choice. Instead of arbitrarily creating a plain inline, we should call `parse_plain` similarly how we do it for `parse_inline` few lines above. Maybe even completely remove check for `next_token.opens()` and call `parse_inline` regardless if token closes or is some content that should simply be taken as is.

Reply:
This is simplified a lot now, thank you for the observation and remarks over video call. The function now only checks if closing token is found, otherwise calls `parse_inline` making this significantly simpler.